### PR TITLE
Use the same context for the specified node in rclcpp::spin functions

### PR DESCRIPTION
**Pull request by HansRobo**
Currently, rclcpp::spin functions, `rclcpp::spin`, `rclcpp::spin_some`, and so on, use the default context.
However they may occur errors when you use the custom context for the specified node of argument of spin functions.

In this pull-request, I set in-place executor to use context of node via executor options.

This is the only fix I found, but there may be other similar fixes.

### CODE REVIEW

**Comment by alsora** [comment]
The approach in the PR looks good to me.

There's several test failures.
Can you take a look?

We should also add a unit-test to verify that the executor is using the correct context

**Review by fujitatomoya — APPROVED** [review]
lgtm, test is ideal.

**Comment by alsora** [comment]
@HansRobo FYI, the feature freeze for Jazzy will be in ~2 weeks, would you be able to add tests before that date?
Otherwise this PR will need to go in the next release.

File: rclcpp/test/rclcpp/executors/test_executors.cpp
Comment (fujitatomoya) [inline rclcpp/test/rclcpp/executors/test_executors.cpp:849]: What we want to test in this particular scope is, these call should work after shutting down the default context to verify those do not rely on the default context? current test is good enough just to makes sure it takes specified node base, but this did not fail before the PR either?

**Comment by fujitatomoya** [comment]
@HansRobo thanks for adding test, i have a question for test.

File: rclcpp/test/rclcpp/executors/test_executors.cpp
Comment (HansRobo) [reply rclcpp/test/rclcpp/executors/test_executors.cpp:849]: This test works regardless of whether the default context is up or down, and it failed before PR as below.

```
Running main() from /opt/ros/rolling/src/gtest_vendor/src/gtest_main.cc
/home/hans/workspace/rclcpp_ws/src/rclcpp/rclcpp/test/rclcpp/executors/test_executors.cpp:852: Failure
Expected: rclcpp::spin_some(node->get_node_base_interface()) doesn't throw an exception.
  Actual: it throws rclcpp::exceptions::RCLInvalidArgument with description "failed to create guard condition: context argument is null, at ./src/rcl/guard_condition.c:65".
/home/hans/workspace/rclcpp_ws/src/rclcpp/rclcpp/test/rclcpp/executors/test_executors.cpp:854: Failure
Expected: rclcpp::spin_all(node->get_node_base_interface(), 1s) doesn't throw an exception.
  Actual: it throws rclcpp::exceptions::RCLInvalidArgument with description "failed to create guard condition: context argument is null, at ./src/rcl/guard_condition.c:65".
/home/hans/workspace/rclcpp_ws/src/rclcpp/rclcpp/test/rclcpp/executors/test_executors.cpp:866: Failure
Expected: check_spin_until_future_complete() doesn't throw an exception.
  Actual: it throws rclcpp::exceptions::RCLInvalidArgument with description "failed to create guard condition: context argument is null, at ./src/rcl/guard_condition.c:65".
```

But as you said, the environment with default context enabled is not suit for this test.
Should I move this test from TestExecutor suite to another or new Suite?
Or, to test in TestExecutor suite, should I shutdown default context in the test? (In this case, perhaps it is necessary to take care not to crash TestExecutors::TearDown)

**Review by fujitatomoya — APPROVED** [review]
lgtm

**Comment by fujitatomoya** [comment]
@alsora since you are interested in this PR. can you take a look at this? i will leave this to you.

**Comment by fujitatomoya** [comment]
https://build.ros2.org/job/Rpr__rclcpp__ubuntu_noble_amd64/106/ is unrelated failure.

**Review by alsora — APPROVED** [review]

**Comment by alsora** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20770/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15068/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21450/)

**Comment by alsora** [comment]
Merging with green full CI (the github action failure is unrelated)

**Comment by HansRobo** [comment]
@fujitatomoya 
I want to use rclcpp with this bug fix in Humble.
Could you backport this to humble? (optionally, iron)

(I'm not a maintainer so I wasn't sure if it was OK to use mergify.)

**Comment by HansRobo** [comment]
@fujitatomoya friendly ping

**Comment by fujitatomoya** [comment]
Mergifyio is not responding, and i see a couple of conflicts need to resolve there.

**Comment by fujitatomoya** [comment]
@HansRobo since the code base difference, i created the iron backport PR with https://github.com/ros2/rclcpp/pull/2618. can you review that?

**Comment by HansRobo** [comment]
@HansRobo  Thanks! I'll review it.
By the way, backporting to Humble is difficult?

**Comment by fujitatomoya** [comment]
@HansRobo i do not think so, once https://github.com/ros2/rclcpp/pull/2618 is merged, we can backport it to humble.

**Comment by fujitatomoya** [comment]
@HansRobo all backports including humble have been completed.

**Comment by HansRobo** [comment]
Thank you!!

**Pull request merged**
